src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java

```diff
@@ -186,14 +191,17 @@ public Bindings getBindings(MethodType mt, FunctionDescriptor cDesc, boolean for
 
     /* Replace DowncallLinker in OpenJDK with the implementation of DowncallLinker specific to OpenJ9 */
     public MethodHandle arrangeDowncall(MethodType mt, FunctionDescriptor cDesc, LinkerOptions options) {
-        // MethodHandle handle = DowncallLinker.getBoundMethodHandle(mt, cDesc, options);
-        // return handle;
-        return null;
+        if (isWinOS) {
+            throw new InternalError("arrangeDowncall is not implemented on Windows/Aarch64");
+        }
+        return DowncallLinker.getBoundMethodHandle(mt, cDesc, options);
     }
 
     /* Replace UpcallLinker in OpenJDK with the implementation of UpcallLinker specific to OpenJ9 */
-    public MemorySegment arrangeUpcall(MethodHandle target, MethodType mt, FunctionDescriptor cDesc, SegmentScope session) {
-        // return UpcallLinker.make(target, mt, cDesc, session);
+    public UpcallStubFactory arrangeUpcall(MethodType mt, FunctionDescriptor cDesc) {
+        if (isWinOS) {
+            throw new InternalError("arrangeUpcall is not implemented on Windows/Aarch64");
+        }
         return null;
```

Comment:
I think this should follow the upstream (b4a1996aa9df668657832975eb12ef1b9411dfd7) code keeping the following instead of `return null;`:
```java
        Bindings bindings = getBindings(mt, cDesc, true);
        final boolean dropReturn = true; /* drop return, since we don't have bindings for it */
        return SharedUtils.arrangeUpcallHelper(mt, bindings.isInMemoryReturn, dropReturn, abiDescriptor(),
                bindings.callingSequence);
```
The same goes for each flavour of `CallArranger.arrangeUpcall()` inherited from openjdk.
That way, uses of `UpcallLinker.makeFactory()` in `SharedUtils.arrangeUpcallHelper()` will not go unnoticed.

Reply:
This is fine in terms of compilation.

Please note we won't directly use this existing code of `SharedUtils.arrangeUpcallHelper()` given it is fundamentally the FFI implementation in OpenJDK, in which the only thing we might care about is just `UpcallLinker.makeFactory()`.

Reply:
To be specific, the signature of `SharedUtils.arrangeUpcallHelper(..) is inappropriate for OpenJ9 as we are only need two arguments `MethodType mt, FunctionDescriptor cDesc) as specified in `arrangeUpcall()`

arrangeUpcallHelper (
   MethodType targetType, <--- required in UpcallLinker in OpenJ9
   boolean isInMemoryReturn, <---- OpenJDK specific & irrelevant  to the FFI code in OpenJ9
    boolean dropReturn,  <---- OpenJDK specific & irrelevant  to the FFI code in OpenJ9
    ABIDescriptor abi,  <---- OpenJDK specific & irrelevant  to the FFI code in OpenJ9
    CallingSequence callingSequence <---- OpenJDK specific & irrelevant  to the FFI code in OpenJ9
  ) {
....
}

So everything related to `Bindings bindings`, `isInMemoryReturn`, `dropReturn`, `abi` and `callingSequence` (which are created based on their own code with `cDesc` in OpenJDK) will be totally removed in our new implementation.

If we have to keep `SharedUtils.arrangeUpcallHelper(..)` in `arrangeUpcall`, it should be something like:
```
    public UpcallStubFactory arrangeUpcall(MethodType mt, FunctionDescriptor cDesc) {
        if (isWinOS) {
            throw new InternalError("arrangeUpcall is not implemented on Windows/Aarch64");
        }
        return SharedUtils.arrangeUpcallHelper(mt, cDesc)`;
}
public static UpcallStubFactory arrangeUpcallHelper(MethodType targetType, FunctionDescriptor cDesc) {
        return UpcallLinker.makeFactory(targetType, cDesc);
    }
```
Or even simply with
```
    public UpcallStubFactory arrangeUpcall(MethodType mt, FunctionDescriptor cDesc) {
        if (isWinOS) {
            throw new InternalError("arrangeUpcall is not implemented on Windows/Aarch64");
        }
        return UpcallLinker.makeFactory(targetType, cDesc);
}
```

Reply:
There's no need to diverge from upstream code if, as I proposed, `UpcallLinker.makeFactory()` (with the upstream signature) throws an exception until we adapt to this new design.